python_modules/dagster/dagster/_core/definitions/op_invocation.py

```diff
@@ -163,14 +169,22 @@ def direct_invocation_result(
                     f"'{context_param_name}', but no value for '{context_param_name}' was "
                     f"found when invoking. Provided kwargs: {kwargs}"
                 )
-            context = cast(UnboundOpExecutionContext, kwargs[context_param_name])
+            if isinstance(kwargs[context_param_name], UnboundAssetExecutionContext):
+                context = cast(UnboundAssetExecutionContext, kwargs[context_param_name])
+            else:
+                context = cast(UnboundOpExecutionContext, kwargs[context_param_name])
             # update kwargs to remove context
             kwargs = {
                 kwarg: val for kwarg, val in kwargs.items() if not kwarg == context_param_name
             }
     # allow passing context, even if the function doesn't have an arg for it
-    elif len(args) > 0 and isinstance(args[0], UnboundOpExecutionContext):
-        context = cast(UnboundOpExecutionContext, args[0])
+    elif len(args) > 0 and isinstance(
+        args[0], (UnboundOpExecutionContext, UnboundAssetExecutionContext)
+    ):
+        if isinstance(args[0], UnboundAssetExecutionContext):
+            context = cast(UnboundAssetExecutionContext, args[0])
+        else:
+            context = cast(UnboundOpExecutionContext, args[0])
```

Comment:
isn't even worth doing all the casting at this point? Seems like the typechecker will be totally confused as to what type "context" is anyways.

From brief inspection it appears the only requirement of the `context` local variable is that it have a `bind` method, so it you really want that typed you can introduce an interface that has this, but this looks like mostly pointless, unnecessary gymnastics to appease typechecking gods, unless I am misreading this